File: test/AdventureWorksLT/SalesOrder.cs
using DevZest.Data;
using DevZest.Data.Annotations;

namespace DevZest.Samples.AdventureWorksLT
{
    public class SalesOrder : SalesOrderHeader
    {
        static SalesOrder()
        {
            RegisterColumn((SalesOrder _) => _.LineCount);
            RegisterChildModel((SalesOrder _) => _.SalesOrderDetails, (SalesOrderDetail _) => _.SalesOrderHeader, null, _ => _.CreateSalesOrderDetail());
        }

        public _Int32 LineCount { get; private set; }

        [ModelValidator]
        private DataValidationError ValidateLineCount(DataRow dataRow)
        {
            return LineCount[dataRow] > 0 ? null : new DataValidationError(UserMessages.Validation_SalesOrder_LineCount, LineCount);
        }

        [Computation(IsAggregate = true)]
        private void ComputeLineCount()
        {
            LineCount.ComputedAs(SalesOrderDetails.SalesOrderDetailID.CountRows());
        }

        public SalesOrderDetail SalesOrderDetails { get; private set; }

        protected virtual SalesOrderDetail CreateSalesOrderDetail()
        {
            return new SalesOrderDetail();
        }

        [Computation(IsAggregate = true)]
        private void ComputeSubTotal()
        {
            SubTotal.ComputedAs(SalesOrderDetails.LineTotal.Sum().IfNull(0), false);
        }
    }
}

File: test/AdventureWorksLT.VB/SalesOrder.vb
Imports DevZest.Data
Imports DevZest.Data.Annotations

Namespace DevZest.Samples.AdventureWorksLT
    Public Class SalesOrder
        Inherits SalesOrderHeader

        Shared Sub New()
            RegisterColumn(Function(x As SalesOrder) x.LineCount)
            RegisterChildModel(Function(x As SalesOrder) x.SalesOrderDetails, Function(x As SalesOrderDetail) x.FK_SalesOrderHeader, Function(x) x.CreateSalesOrderDetail())
        End Sub

        Private m_LineCount As _Int32
        Public Property LineCount As _Int32
            Get
                Return m_LineCount
            End Get
            Private Set
                m_LineCount = Value
            End Set
        End Property

        <ModelValidator>
        Private Function ValidateLineCount(ByVal dataRow As DataRow) As DataValidationError
            Return If(LineCount(dataRow) > 0, Nothing, New DataValidationError(UserMessages.Validation_SalesOrder_LineCount, LineCount))
        End Function

        <Computation(ComputationMode.Aggregate)>
        Private Sub ComputeLineCount()
            LineCount.ComputedAs(SalesOrderDetails.SalesOrderDetailID.CountRows())
        End Sub

        Private m_SalesOrderDetails As SalesOrderDetail
        Public Property SalesOrderDetails As SalesOrderDetail
            Get
                Return m_SalesOrderDetails
            End Get
            Private Set
                m_SalesOrderDetails = Value
            End Set
        End Property

        Protected Overridable Function CreateSalesOrderDetail() As SalesOrderDetail
            Return New SalesOrderDetail()
        End Function

        <Computation(ComputationMode.Aggregate)>
        Private Sub ComputeSubTotal()
            SubTotal.ComputedAs(SalesOrderDetails.LineTotal.Sum().IfNull(0), False)
        End Sub
    End Class
End Namespace
